datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -441,6 +491,62 @@ impl CaseExpr {
 
         Ok(ColumnarValue::Array(zip(&remainder, &else_, &then_value)?))
     }
+
+    /// This is a specialization that fuses a greater-than-zero guard with division so
+    /// so don't need to be lazy. This is a very common guard in TPC-DS as well as normal
+    /// human expressions.
+    ///
+    /// `CASE WHEN y > 0 THEN x / y ELSE NULL END`
+    ///
+    /// It is executed as though `x / NULL_IF(y <= 0, y)` which is safe because arrow-rs
+    /// skips evaluating fallible functions if one of their inputs is null.
+    fn div_gt_0(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
```

Comment:
If y'all like this path I'm happy to continue down it, adding the missing tests and removing the rough edges. I'm certainly plugging things in wrong, though the one test case we have now is passing.